File: calculators/noise_exposure.py
"""
Noise Exposure & Daily Noise Dose Calculator
Author: Diego Malpica

Usage:
    Provides functions to calculate the percentage noise dose and permissible
    exposure durations according to the U.S. OSHA occupational noise standard
    (29 CFR §1910.95) and the NIOSH Recommended Exposure Limit (REL) for noise.
    The calculator is intended for occupational and industrial hygiene studies
    and *must not* be used as a legal compliance tool without proper sound-level
    measurements and professional judgement.

Scientific & Regulatory Basis (APA style):
    Occupational Safety and Health Administration. (2023). *29 CFR §1910.95 –
    Occupational noise exposure.* https://www.osha.gov/laws-regs/regulations/

    National Institute for Occupational Safety and Health. (1998). *Criteria for
    a recommended standard: Occupational noise exposure* (DHHS Publication
    No. 98-126). U.S. Department of Health and Human Services.

    Suter, A. H. (1991). *Noise and its effects.* Administrative Conference of
    the United States.
"""
from __future__ import annotations

__all__ = [
    "permissible_duration",
    "noise_dose_osha",
    "noise_dose_niosh",
]

# ----------------------------------------------------------------------------
# Core Functions
# ----------------------------------------------------------------------------

def permissible_duration(
    spl_dba: float,
    *,
    criterion_level: float = 90.0,
    exchange_rate: float = 5.0,
) -> float:
    """Return permissible exposure duration in *hours* at a given SPL.

    Parameters
    ----------
    spl_dba : float
        A-weighted sound pressure level, dB(A).
    criterion_level : float, default 90 dBA
        Criterion level (Lc). OSHA uses 90 dBA. NIOSH uses 85 dBA.
    exchange_rate : float, default 5 dB
        Exchange rate (q). OSHA uses 5 dB. NIOSH uses 3 dB.

    Notes
    -----
    Formula (OSHA, NIOSH):
        T = 8 h / 2^((SPL − Lc) / q)
    """
    exponent = (spl_dba - criterion_level) / exchange_rate
    T_hours = 8.0 / (2 ** exponent)
    return T_hours


def _dose_percent(spl_dba: float, duration_hours: float,
                  *, criterion_level: float, exchange_rate: float) -> float:
    """Internal helper that computes percent dose for a single SPL block."""
    T_allowed = permissible_duration(spl_dba,
                                     criterion_level=criterion_level,
                                     exchange_rate=exchange_rate)
    dose_pct = 100.0 * (duration_hours / T_allowed)
    return dose_pct


def noise_dose_osha(levels: list[float], durations_hours: list[float]) -> float:
    """Compute total OSHA noise dose (percent) for multiple exposure segments.

    Parameters
    ----------
    levels : list[float]
        List of A-weighted sound levels in dB(A).
    durations_hours : list[float]
        Corresponding exposure durations in hours.

    Returns
    -------
    float
        Total daily dose in percent. ≥100 % exceeds the permissible dose.
    """
    if len(levels) != len(durations_hours):
        raise ValueError("`levels` and `durations_hours` length mismatch.")

    total_dose = 0.0
    for L, t in zip(levels, durations_hours):
        total_dose += _dose_percent(L, t, criterion_level=90.0, exchange_rate=5.0)
    return total_dose


def noise_dose_niosh(levels: list[float], durations_hours: list[float]) -> float:
    """Compute total NIOSH REL noise dose (percent).

    NIOSH uses criterion level 85 dBA and 3-dB exchange rate.
    """
    if len(levels) != len(durations_hours):
        raise ValueError("`levels` and `durations_hours` length mismatch.")

    total_dose = 0.0
    for L, t in zip(levels, durations_hours):
        total_dose += _dose_percent(L, t, criterion_level=85.0, exchange_rate=3.0)
    return total_dose

File: frontend/src/calculators/noiseExposure.ts
/**
 * Noise Exposure & Daily Noise Dose
 *
 * OSHA (29 CFR §1910.95): criterion 90 dBA, exchange rate 5 dB.
 * NIOSH REL (DHHS Pub. No. 98-126): criterion 85 dBA, exchange rate 3 dB.
 *
 * For occupational and industrial-hygiene research/education only — not a
 * legal compliance tool.
 *
 * References:
 *   - OSHA 29 CFR §1910.95 (2023). https://www.osha.gov/laws-regs/regulations/
 *   - NIOSH (1998). DHHS Publication 98-126.
 *   - Suter, A. H. (1991). Noise and its effects.
 */

/**
 * Permissible exposure duration (hours) at a given SPL.
 *
 * T = 8 / 2^((SPL − Lc) / q)
 *
 * @param spl_dba Sound pressure level, dB(A)
 * @param criterion_level Lc — criterion level (default 90 dBA, OSHA)
 * @param exchange_rate q — exchange rate (default 5 dB, OSHA)
 */
export function permissibleDuration(
  spl_dba: number,
  criterion_level: number = 90.0,
  exchange_rate: number = 5.0
): number {
  const exponent = (spl_dba - criterion_level) / exchange_rate;
  return 8.0 / Math.pow(2, exponent);
}

function dosePercent(
  spl_dba: number,
  duration_hours: number,
  criterion_level: number,
  exchange_rate: number
): number {
  const allowed = permissibleDuration(spl_dba, criterion_level, exchange_rate);
  return 100 * (duration_hours / allowed);
}

/**
 * Total OSHA noise dose (%) summed across multiple exposure segments.
 * ≥ 100% exceeds the permissible daily dose.
 */
export function noiseDoseOsha(levels_dba: number[], durations_hours: number[]): number {
  if (levels_dba.length !== durations_hours.length) {
    throw new Error('`levels_dba` and `durations_hours` length mismatch');
  }
  let total = 0;
  for (let i = 0; i < levels_dba.length; i++) {
    total += dosePercent(levels_dba[i], durations_hours[i], 90.0, 5.0);
  }
  return total;
}

/**
 * Total NIOSH REL noise dose (%) — criterion 85 dBA, exchange rate 3 dB.
 */
export function noiseDoseNiosh(levels_dba: number[], durations_hours: number[]): number {
  if (levels_dba.length !== durations_hours.length) {
    throw new Error('`levels_dba` and `durations_hours` length mismatch');
  }
  let total = 0;
  for (let i = 0; i < levels_dba.length; i++) {
    total += dosePercent(levels_dba[i], durations_hours[i], 85.0, 3.0);
  }
  return total;
}
